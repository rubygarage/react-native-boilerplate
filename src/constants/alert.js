// eslint-disable-next-line import/prefer-default-export
export const ALERT_TYPES = {
  error: 'error',
  success: 'success',
  info: 'info',
};
