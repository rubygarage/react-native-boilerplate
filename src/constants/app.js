export const GLOBAL_ERROR = '_error';
export const OFFLINE_FLASH_ID = 'offline';
export const OFFLINE_FLASH_PRIORITY = 0;
export const FLASH_LIFETIME = 3000;

export const PLATFORM_TYPE = {
  android: 'android',
  ios: 'ios',
};
