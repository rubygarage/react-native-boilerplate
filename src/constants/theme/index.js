export default {
  colors: {
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
  },
  sizes: {
  },
};
