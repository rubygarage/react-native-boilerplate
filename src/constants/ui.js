// eslint-disable-next-line import/prefer-default-export
export const POPOVER_PLACEMENT = {
  bottom: 'bottom',
  auto: 'auto',
};
