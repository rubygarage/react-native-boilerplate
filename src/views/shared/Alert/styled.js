import styled from 'styled-components/native';

import Icon from 'views/shared/Icon';

const Wrapper = styled.View`
  position: relative;
  flex-direction: row;
  margin-bottom: 8;
  padding-top: 12;
  padding-right: ${({ onCloseClick }) => (onCloseClick ? 48 : 12)};
  padding-bottom: 12;
  padding-left: 12;
  border-radius: 4;
  background-color: ${({ color }) => color};
`;

const IconWrap = styled.View`
  margin-right: 12;
  flex-shrink: 0;
`;

const ContentArea = styled.View`
  flex-grow: 1;
  padding-right: 36;
  display: flex;
  align-items: flex-start;
`;

const MessageTitle = styled.Text`
  flex-direction: row;
  flex-wrap: wrap;
  margin-top: 2;
  font-size: 14;
  line-height: 20;
  font-weight: bold;
  color: ${({ color }) => color};
`;

const MessageText = styled.Text`
  margin-bottom: 2;
  font-size: 14;
  line-height: 20;
  color: ${({ color }) => color};
`;

const ActionButton = styled.TouchableOpacity`
  flex-direction: row;
  align-items: center;
  max-width: 200;
  height: 36;
  margin-top: 8;
  padding-right: 12;
  padding-left: 12;
  border-radius: 4;
  background-color: ${({ color }) => color};
`;

const ActionIcon = styled(Icon)`
  margin-right: 24;
`;

const ActionButtonText = styled.Text`
  color: ${({ theme }) => theme.colors.white};
  font-size: 14;
  line-height: 20;
  font-weight: bold;
  letter-spacing: -0.2;
`;

const CloseButton = styled.TouchableOpacity`
  position: absolute;
  top: 12;
  right: 12;
  color: ${({ theme }) => theme.colors.white};
`;

export {
  Wrapper,
  IconWrap,
  ContentArea,
  MessageTitle,
  MessageText,
  ActionButton,
  ActionButtonText,
  CloseButton,
  ActionIcon,
};
