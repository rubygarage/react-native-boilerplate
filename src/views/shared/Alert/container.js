import React, { Component } from 'react';
import { withTheme } from 'styled-components/native';
import PropTypes from 'prop-types';

import { ALERT_TYPES } from 'constants/alert';
import AlertComponent from './component';

class Alert extends Component {
  getStyleProps = () => {
    const { type, theme } = this.props;
    switch (type) {
      case ALERT_TYPES.error:
        return { iconName: 'close', contentColor: theme.colors.red500, backgroundColor: theme.colors.red50 };
      case ALERT_TYPES.success:
        return { iconName: 'tick', contentColor: theme.colors.green500, backgroundColor: theme.colors.green50 };
      default:
        return { iconName: 'info', contentColor: theme.colors.blue500, backgroundColor: theme.colors.blue50 };
    }
  }

  render() {
    const styleProps = this.getStyleProps();
    return (
      <AlertComponent
        {...this.props}
        {...styleProps}
      />
    );
  }
}

Alert.propTypes = {
  type: PropTypes.string.isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape().isRequired }).isRequired,
};

export { Alert as AlertContainer };
export default withTheme(Alert);
