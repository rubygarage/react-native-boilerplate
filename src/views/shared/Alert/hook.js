import { useContext } from 'react';
import { ThemeContext } from 'styled-components';

import { ALERT_TYPES } from 'constants/alert';

const useContainer = () => {
  const theme = useContext(ThemeContext);

  const getStyleProps = (type) => {
    switch (type) {
      case ALERT_TYPES.error:
        return { iconName: 'close', contentColor: theme.colors.red500, backgroundColor: theme.colors.red50 };
      case ALERT_TYPES.success:
        return { iconName: 'tick', contentColor: theme.colors.green500, backgroundColor: theme.colors.green50 };
      default:
        return { iconName: 'info', contentColor: theme.colors.blue500, backgroundColor: theme.colors.blue50 };
    }
  };

  return {
    theme,
    getStyleProps,
  };
};

export default useContainer;
