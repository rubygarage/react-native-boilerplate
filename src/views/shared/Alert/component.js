import React, { useContext } from 'react';
import { Text } from 'react-native';
import PropTypes from 'prop-types';
import { is } from 'ramda';
import { FormattedMessage } from 'react-intl';

import { ThemeContext } from 'styled-components';

import { ALERT_TYPES } from 'constants/alert';
import Icon from 'views/shared/Icon';
import * as S from './styled';

export const getStyleProps = (type, theme) => {
  switch (type) {
    case ALERT_TYPES.error:
      return { iconName: 'close', contentColor: theme.colors.red500, backgroundColor: theme.colors.red50 };
    case ALERT_TYPES.success:
      return { iconName: 'tick', contentColor: theme.colors.green500, backgroundColor: theme.colors.green50 };
    default:
      return { iconName: 'info', contentColor: theme.colors.blue500, backgroundColor: theme.colors.blue50 };
  }
};

const Alert = ({
  title,
  subtitle,
  message,
  onCloseClick,
  onActionClick,
  actionIconName,
  actionTextId,
  type,
}) => {
  const theme = useContext(ThemeContext);

  const {
    backgroundColor,
    iconName,
    contentColor,
  } = getStyleProps(type, theme);

  return (
    <S.Wrapper color={backgroundColor} onCloseClick={onCloseClick}>
      <S.IconWrap>
        <Icon name={iconName} size={28} color={contentColor} />
      </S.IconWrap>
      <S.ContentArea>
        <Text>
          <S.MessageTitle color={contentColor}>
            {is(Object, title) ? <FormattedMessage id={title.id} /> : title}
          </S.MessageTitle>
          {subtitle && (
          <S.MessageText color={contentColor}>
            {is(Object, subtitle) ? <FormattedMessage id={subtitle.id} /> : subtitle}
          </S.MessageText>
          )}
        </Text>
        {message && (
        <S.MessageText color={contentColor}>
          {is(Object, message) ? <FormattedMessage id={message.id} /> : message}
        </S.MessageText>
        )}
        {onActionClick && actionTextId
        && (
          <S.ActionButton color={contentColor} onPress={onActionClick}>
            {actionIconName && (
              <S.ActionIcon name={actionIconName} size={28} color={theme.colors.white} />
            )}
            <S.ActionButtonText>
              <FormattedMessage id={actionTextId} />
            </S.ActionButtonText>
          </S.ActionButton>
        )}
      </S.ContentArea>
      {onCloseClick
      && (
        <S.CloseButton onPress={onCloseClick}>
          <Icon name="cross" size={28} color={contentColor} />
        </S.CloseButton>
      )}
    </S.Wrapper>
  );
};

Alert.propTypes = {
  title: PropTypes.oneOfType([PropTypes.shape(), PropTypes.string]).isRequired,
  subtitle: PropTypes.oneOfType([PropTypes.shape(), PropTypes.string]),
  message: PropTypes.oneOfType([PropTypes.shape(), PropTypes.string]),
  type: PropTypes.string.isRequired,
  onCloseClick: PropTypes.func,
  onActionClick: PropTypes.func,
  actionIconName: PropTypes.string,
  actionTextId: PropTypes.string,
};

Alert.defaultProps = {
  subtitle: undefined,
  message: undefined,
  onCloseClick: undefined,
  onActionClick: undefined,
  actionIconName: undefined,
  actionTextId: undefined,
};

export default Alert;
