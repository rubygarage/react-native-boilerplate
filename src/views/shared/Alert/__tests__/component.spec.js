import * as React from 'react';
import { shallow } from 'enzyme';

import { ALERT_TYPES } from 'constants/alert';
import theme from 'utils/testHelpers/mockedTheme';
import testIntl from 'utils/testHelpers/testIntl';
import HooksTestHelper from 'utils/testHelpers/hooksTestHelper';

import Alert, { getStyleProps } from '../component';

const defaultProps = {
  title: { id: 'title_id' },
  subtitle: { id: 'subtitle_id' },
  message: { id: 'message_id' },
  intl: testIntl,
  type: 'info',
};

describe('Alert component', () => {
  HooksTestHelper.mockUseContextImplementation(() => theme);

  const component = shallow(<Alert {...defaultProps} />);

  it('renders correctly default', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with string data', () => {
    const props = {
      ...defaultProps,
      title: 'test title',
      subtitle: 'test title',
      message: 'test message',
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly without subtitle', () => {
    const props = {
      ...defaultProps,
      subtitle: undefined,
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly closable', () => {
    const props = {
      ...defaultProps,
      onCloseClick: jest.fn(),
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with action button', () => {
    const props = {
      ...defaultProps,
      onActionClick: jest.fn(),
      actionIconName: 'video',
      actionTextId: 'testActionTextId',
    };
    component.setProps(props);
    expect(component).toMatchSnapshot();
  });

  describe('getStyleProps', () => {
    it('returns info style props', () => {
      expect(getStyleProps(ALERT_TYPES.info, theme)).toEqual({
        contentColor: theme.colors.blue500,
        backgroundColor: theme.colors.blue50,
        iconName: 'info',
      });
    });

    it('returns error style props', () => {
      expect(getStyleProps(ALERT_TYPES.error, theme)).toEqual({
        contentColor: theme.colors.red500,
        backgroundColor: theme.colors.red50,
        iconName: 'close',
      });
    });

    it('returns success style props', () => {
      expect(getStyleProps(ALERT_TYPES.success, theme)).toEqual({
        contentColor: theme.colors.green500,
        backgroundColor: theme.colors.green50,
        iconName: 'tick',
      });
    });
  });
});
