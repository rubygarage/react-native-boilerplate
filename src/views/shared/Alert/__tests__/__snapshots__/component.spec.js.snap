// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert component renders correctly closable 1`] = `
<Styled(View)
  color="[mocked blue50]"
  onCloseClick={[MockFunction]}
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="info"
      size={28}
    />
  </Styled(View)>
  <Styled(View)>
    <Text>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="title_id"
        />
      </Styled(Text)>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="subtitle_id"
        />
      </Styled(Text)>
    </Text>
    <Styled(Text)
      color="[mocked blue500]"
    >
      <FormattedMessage
        id="message_id"
      />
    </Styled(Text)>
  </Styled(View)>
  <Styled(Component)
    onPress={[MockFunction]}
  >
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="cross"
      size={28}
    />
  </Styled(Component)>
</Styled(View)>
`;

exports[`Alert component renders correctly default 1`] = `
<Styled(View)
  color="[mocked blue50]"
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="info"
      size={28}
    />
  </Styled(View)>
  <Styled(View)>
    <Text>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="title_id"
        />
      </Styled(Text)>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="subtitle_id"
        />
      </Styled(Text)>
    </Text>
    <Styled(Text)
      color="[mocked blue500]"
    >
      <FormattedMessage
        id="message_id"
      />
    </Styled(Text)>
  </Styled(View)>
</Styled(View)>
`;

exports[`Alert component renders correctly with action button 1`] = `
<Styled(View)
  color="[mocked blue50]"
  onCloseClick={[MockFunction]}
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="info"
      size={28}
    />
  </Styled(View)>
  <Styled(View)>
    <Text>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="title_id"
        />
      </Styled(Text)>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="subtitle_id"
        />
      </Styled(Text)>
    </Text>
    <Styled(Text)
      color="[mocked blue500]"
    >
      <FormattedMessage
        id="message_id"
      />
    </Styled(Text)>
    <Styled(Component)
      color="[mocked blue500]"
      onPress={[MockFunction]}
    >
      <Styled(Icon)
        color="[mocked white]"
        name="video"
        size={28}
      />
      <Styled(Text)>
        <FormattedMessage
          id="testActionTextId"
        />
      </Styled(Text)>
    </Styled(Component)>
  </Styled(View)>
  <Styled(Component)
    onPress={[MockFunction]}
  >
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="cross"
      size={28}
    />
  </Styled(Component)>
</Styled(View)>
`;

exports[`Alert component renders correctly with string data 1`] = `
<Styled(View)
  color="[mocked blue50]"
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="info"
      size={28}
    />
  </Styled(View)>
  <Styled(View)>
    <Text>
      <Styled(Text)
        color="[mocked blue500]"
      >
        test title
      </Styled(Text)>
      <Styled(Text)
        color="[mocked blue500]"
      >
        test title
      </Styled(Text)>
    </Text>
    <Styled(Text)
      color="[mocked blue500]"
    >
      test message
    </Styled(Text)>
  </Styled(View)>
</Styled(View)>
`;

exports[`Alert component renders correctly without subtitle 1`] = `
<Styled(View)
  color="[mocked blue50]"
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="[mocked blue500]"
      name="info"
      size={28}
    />
  </Styled(View)>
  <Styled(View)>
    <Text>
      <Styled(Text)
        color="[mocked blue500]"
      >
        <FormattedMessage
          id="title_id"
        />
      </Styled(Text)>
    </Text>
    <Styled(Text)
      color="[mocked blue500]"
    >
      <FormattedMessage
        id="message_id"
      />
    </Styled(Text)>
  </Styled(View)>
</Styled(View)>
`;
